src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.cs

```diff
@@ -279,6 +278,7 @@ void RemoveEnv()
 
             private void MakeAndAssignEnvironments()
             {
+                var closuresCapturingScopeVariables = new Dictionary<Scope, HashSet<Closure>>();
```

Comment:
Visiting the scope tree is meant to be pretty cheap, compared to visiting the syntax tree or the bound node tree. Rather than record this information during an unrelated pass I'd re-visit the scope tree in your merge function.

Reply:
We should also use PooledDictionary and PooledHashSet